extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/ReadMongoP.java

```diff
@@ -0,0 +1,482 @@
+            while ((doc = delegate.next()) != null) {
+                chunk.add(mapItemFn.apply(doc));
+                lastItem = doc;
+            }
+            lastKey = lastItem == null ? null : lastItem.getObjectId("__id");
+
+            return Traversers.traverseIterable(chunk);
+        }
+
+        @Override
+        boolean everCompletes() {
+            return true;
+        }
+
+        @Nonnull
+        @Override
+        public Object snapshot() {
+            return lastKey;
+        }
+
+        @Override
+        public void restore(Object value) {
+            lastKey = (ObjectId) value;
+        }
+
+        @Override
+        public void close() {
+
+        }
+    }
+
+    private final class StreamMongoReader extends MongoChunkedReader {
+        private final FunctionEx<ChangeStreamDocument<Document>, I> mapFn;
+        private final Long startTimestamp;
+        private final List<Bson> aggregates;
+        private final EventTimeMapper<I> eventTimeMapper;
+        private MongoCursor<ChangeStreamDocument<Document>> cursor;
+        private BsonDocument resumeToken;
+
+        private StreamMongoReader(
+                String databaseName,
+                String collectionName,
+                FunctionEx<ChangeStreamDocument<Document>, I> mapFn,
+                Long startTimestamp,
+                List<Bson> aggregates,
+                EventTimeMapper<I> eventTimeMapper
+        ) {
+            super(databaseName, collectionName);
+            this.mapFn = mapFn;
+            this.startTimestamp = startTimestamp;
+            this.aggregates = aggregates;
+            this.eventTimeMapper = eventTimeMapper;
+        }
+
+        @Override
+        public void onConnect(MongoClient mongoClient, boolean snapshotsEnabled) {
+            List<Bson> aggregateList = new ArrayList<>(aggregates);
+            if (totalParallelism > 1) {
+                aggregateList.addAll(0, partitionAggregate(totalParallelism, processorIndex, true));
+            }
+            ChangeStreamIterable<Document>  changeStream;
+            if (collection != null) {
+                changeStream = collection.watch(aggregateList);
+            } else if (database != null) {
+                changeStream = database.watch(aggregateList);
+            } else {
+                changeStream = mongoClient.watch(aggregateList);
+            }
+
+            if (resumeToken != null) {
+                changeStream.resumeAfter(resumeToken);
+            } else if (startTimestamp != null) {
+                changeStream.startAtOperationTime(new BsonTimestamp(startTimestamp));
+            }
+            cursor = changeStream.batchSize(BATCH_SIZE).iterator();
+        }
+
+        @Override
+        boolean everCompletes() {
+            return false;
+        }
+
+        @Nonnull
+        @Override
+        public Traverser<?> nextChunkTraverser() {
+            try {
+                List<ChangeStreamDocument<Document>> chunk = readChunk();
```

Comment:
Again, I think we could avoid the list here.

Reply:
When I change it, `testStream_whenWatchAll` hangs ;) 

This is also why I used `tryNext` instead of `hasNext` + `next()` - `tryNext` does not block, if there isn't anything in the cursor it just returns null.